shell/platform/darwin/ios/framework/Source/FlutterEngine.mm

```diff
@@ -493,22 +506,43 @@ - (BOOL)createShell:(NSString*)entrypoint libraryURI:(NSString*)libraryURI {
     FML_LOG(ERROR) << "Could not start a shell FlutterEngine with entrypoint: "
                    << entrypoint.UTF8String;
   } else {
-    [self setupChannels];
-    if (!_platformViewsController) {
-      _platformViewsController.reset(new flutter::FlutterPlatformViewsController());
+    if (!oc_async_init_block) {
+      [self setUpWhenInit];
     }
-    _publisher.reset([[FlutterObservatoryPublisher alloc] init]);
-    [self maybeSetupPlatformViewChannels];
-    _shell->GetIsGpuDisabledSyncSwitch()->SetSwitch(_isGpuDisabled ? true : false);
   }
-
   return _shell != nullptr;
 }
 
+- (BOOL)createShell:(NSString*)entrypoint libraryURI:(NSString*)libraryURI {
+  return [self createShell:entrypoint libraryURI:libraryURI asyncInitCallback:nil];
+}
+
+- (void)setUpWhenInit {
+  [self setupChannels];
+  if (!_platformViewsController) {
+    _platformViewsController.reset(new flutter::FlutterPlatformViewsController());
+  }
+  _publisher.reset([[FlutterObservatoryPublisher alloc] init]);
+  [self maybeSetupPlatformViewChannels];
+  _shell->GetIsGpuDisabledSyncSwitch()->SetSwitch(_isGpuDisabled ? true : false);
+}
+
 - (BOOL)run {
   return [self runWithEntrypoint:FlutterDefaultDartEntrypoint libraryURI:nil];
 }
 
+- (void)asyncRun:(InitCallBackBlock)asyncCallback {
+  asyncCallback = [asyncCallback copy];
+  [self createShell:FlutterDefaultDartEntrypoint
+             libraryURI:nil
+      asyncInitCallback:^() {
+        [self setUpWhenInit];
+        [self launchEngine:FlutterDefaultDartEntrypoint libraryURI:nil];
+        asyncCallback();
```

Comment:
Would crash when asyncCallback is nil.

Reply:
If asyncCallback is null .then let it crash when dev.
This api design for async. callback must be not null.  if is null ,develop will never know when engine init end . if they call other flutter api before engien init , will make other strange error

Reply:
Can't agree with this.
Even so, you're supposed to add an assert expression.

Besides, what if the parameter is nonnull in development while be passed in with an null parameter after release. I don't think it is good to handle in that way.

Reply:
done. 
- on debug mode  : add assert
- on release mode : add NP protection